fix(curriculum): replace regex-based tests with Explorer AST helper in Fortune Teller App Workshop (#69081)
freeCodeCamp/freeCodeCamp: curriculum/challenges/english/blocks/workshop-fortune-teller-app/68f1e8de2731b07d62e738d2.md

```js
const explorer = await __helpers.Explorer(code);
const { showFortune } = explorer.classes.Game?.methods ?? {};
const { foundCard } = showFortune?.variables ?? {};
assert.exists(foundCard);
```

> Your `foundCard` should be assigned by calling `find` on `this.cards`, passing a callback that takes a `card` parameter and implicitly returns the result of comparing `card.name_short` with `cardId` using strict equality.

Reviewer: ```suggestion
Your `foundCard` should be assigned by calling `find` on `this.cards`, passing a callback that takes a `card` parameter and returns the result of comparing `card.name_short` with `cardId` using strict equality.
```

```js
const explorer = await __helpers.Explorer(code);
const { showFortune } = explorer.classes.Game?.methods ?? {};
const { foundCard } = showFortune?.variables ?? {};

const validForms = [
  'this.cards.find((card) => card.name_short === cardId)',
  'this.cards.find((card) => { return card.name_short === cardId; })',
  'this.cards.find((card) => cardId === card.name_short)',
  'this.cards.find((card) => { return cardId === card.name_short; })'

> ];

Reviewer 2: This hint currently fails for solutions where `card` isn't wrapped in parentheses. Please update the logic to account for those cases as well.

```js
const explorer = await __helpers.Explorer(code);
const { showFortune } = explorer.classes.Game?.methods ?? {};
const { foundCard } = showFortune?.variables ?? {};
assert.exists(foundCard);
```

Your `foundCard` should be assigned by calling `find` on `this.cards`, passing a callback that takes a `card` parameter and returns the result of comparing `card.name_short` with `cardId` using strict equality.

```js
const explorer = await __helpers.Explorer(code);
const { showFortune } = explorer.classes.Game?.methods ?? {};
const { foundCard } = showFortune?.variables ?? {};

const validForms = [
  'this.cards.find((card) => card.name_short === cardId)',
  'this.cards.find((card) => { return card.name_short === cardId; })',
  'this.cards.find((card) => cardId === card.name_short)',
  'this.cards.find((card) => { return cardId === card.name_short; })',
  'this.cards.find(card => card.name_short === cardId)',
  'this.cards.find(card => { return card.name_short === cardId; })',
  'this.cards.find(card => cardId === card.name_short)',

>   'this.cards.find(card => { return cardId === card.name_short; })'

Reviewer: ```suggestion
  'this.cards.find((card) => card.name_short === cardId)',
  'this.cards.find((card) => { return card.name_short === cardId; })',
  'this.cards.find((card) => cardId === card.name_short)',
  'this.cards.find((card) => { return cardId === card.name_short; })',
  'this.cards.find(card => card.name_short === cardId)',
  'this.cards.find(card => { return card.name_short === cardId; })',
  'this.cards.find(card => cardId === card.name_short)',
  'this.cards.find(card => { return cardId === card.name_short; })',
  'this.cards.find((card: Card) => card.name_short === cardId)',
  'this.cards.find((card: Card) => { return card.name_short === cardId; })',
  'this.cards.find((card: Card) => cardId === card.name_short)',
  'this.cards.find((card: Card) => { return cardId === card.name_short; })'
```

assert.exists(foundCard);
```

Your `foundCard` should be assigned by calling `find` on `this.cards`, passing a callback that takes a `card` parameter and returns the result of comparing `card.name_short` with `cardId` using strict equality.

```js
const explorer = await __helpers.Explorer(code);
const { showFortune } = explorer.classes.Game?.methods ?? {};
const { foundCard } = showFortune?.variables ?? {};

const validForms = [
  'this.cards.find((card) => card.name_short === cardId)',
  'this.cards.find((card) => { return card.name_short === cardId; })',
  'this.cards.find((card) => cardId === card.name_short)',
  'this.cards.find((card) => { return cardId === card.name_short; })',
  'this.cards.find(card => card.name_short === cardId)',
  'this.cards.find(card => { return card.name_short === cardId; })',
  'this.cards.find(card => cardId === card.name_short)',
  'this.cards.find(card => { return cardId === card.name_short; })',
  'this.cards.find((card: Card) => card.name_short === cardId)',
  'this.cards.find((card: Card) => { return card.name_short === cardId; })',
  'this.cards.find((card: Card) => cardId === card.name_short)',

>   'this.cards.find((card: Card) => { return cardId === card.name_short; })'

Reviewer: ```suggestion
  'this.cards.find((card: Card) => { return cardId === card.name_short; })',
  'this.cards.find(function (card) { return card.name_short === cardId; })',
  'this.cards.find(function (card) { return cardId === card.name_short; })',
  'this.cards.find(function (card: Card) { return card.name_short === cardId; })',
  'this.cards.find(function (card: Card) { return cardId === card.name_short; })'
```

```js
const explorer = await __helpers.Explorer(code);

> assert.exists(explorer.classes.Game?.methods?.showFortune?.variables?.foundCard);

Reviewer 2: Please avoid long chains, it worsens the readability of the code, so use destructuring instead.

```js
const explorer = await __helpers.Explorer(code);
const { showFortune } = explorer.classes.Game?.methods ?? {};
const { foundCard } = showFortune?.variables ?? {};
assert.exists(foundCard);
```

> Your `foundCard` should be assigned by calling `find` on `this.cards` with the correct callback.

Reviewer 2: ```suggestion
Your `foundCard` should be assigned by calling `find` on `this.cards`, passing a callback that takes a `card` parameter and implicitly returns the result of comparing `card.name_short` with `cardId` using strict equality.
```

```js
const explorer = await __helpers.Explorer(code);
const { showFortune } = explorer.classes.Game?.methods ?? {};
const { foundCard } = showFortune?.variables ?? {};
assert.isTrue(foundCard?.value.matches('this.cards.find((card) => card.name_short === cardId)'));
```

The `find` callback should have a parameter named `card`.

```js
const explorer = await __helpers.Explorer(code);
const { showFortune } = explorer.classes.Game?.methods ?? {};
const { foundCard } = showFortune?.variables ?? {};
const expression = foundCard?.value.toString();
assert.match(expression, /cards\.find\s*\(\s*\(\s*card\s*\)\s*=>/);
```

Inside the callback, `card.name_short` should be compared with `cardId`.

```js
const explorer = await __helpers.Explorer(code);
const { showFortune } = explorer.classes.Game?.methods ?? {};
const { foundCard } = showFortune?.variables ?? {};
const expression = foundCard?.value.toString();
assert.match(expression, /card\.name_short\s*===?\s*cardId/);

> ```

Reviewer 2: We can get rid of these hints, as the previous hint is sufficient.
